src/sentry/snuba/metrics.py

```diff
@@ -0,0 +1,329 @@
+    # use the materialized storage due to no grouping on the `started` column.
+    # NOTE: we can remove the difference between `interval` / `rounding_interval`
+    # as soon as snuba can provide us with grouped totals in the same query
+    # as the timeseries (using `WITH ROLLUP` in clickhouse)
+    rounding_interval = int(math.ceil(interval / ONE_HOUR) * ONE_HOUR)
+    date_range = timedelta(
+        seconds=int(rounding_interval * math.ceil(date_range.total_seconds() / rounding_interval))
+    )
+
+    # TODO: restrictions
+
+    end_ts = int(rounding_interval * math.ceil(to_timestamp(end) / rounding_interval))
+    end = to_datetime(end_ts)
+    # when expanding the rounding interval, we would adjust the end time too far
+    # to the future, in which case the start time would not actually contain our
+    # desired date range. adjust for this by extend the time by another interval.
+    # for example, when "45m" means the range from 08:49:00-09:34:00, our rounding
+    # has to go from 08:00:00 to 10:00:00.
+    if rounding_interval > interval and (end - date_range) > start:
+        date_range += timedelta(seconds=rounding_interval)
+    start = end - date_range
+
+    # snuba <-> sentry has a 5 minute cache for *exact* queries, which these
+    # are because of the way we do our rounding. For that reason we round the end
+    # of "realtime" queries to one minute into the future to get a one-minute cache instead.
+    if end > now:
+        end = to_datetime(ONE_MINUTE * (math.floor(to_timestamp(now) / ONE_MINUTE) + 1))
+
+    return start, end, interval
+
+
+class QueryDefinition:
+    """
+    This is the definition of the query the user wants to execute.
+    This is constructed out of the request params, and also contains a list of
+    `fields` and `groupby` definitions as [`ColumnDefinition`] objects.
+
+
+    Adapted from [`sentry.snuba.sessions_v2`].
+
+    """
+
+    def __init__(self, query_params, allow_minute_resolution=False):
+        self.query = query_params.get("query", "")
+        raw_fields = query_params.getlist("field", [])
+        self.groupby = query_params.getlist("groupBy", [])
+
+        if len(raw_fields) == 0:
+            raise InvalidField('Request is missing a "field"')
+
+        self.fields = {key: parse_field(key) for key in raw_fields}
+
+        start, end, rollup = get_constrained_date_range(query_params, allow_minute_resolution)
+        self.rollup = rollup
+        self.start = start
+        self.end = end
+
+    def get_intervals(self):
+        start = self.start
+        end = self.end
+        delta = timedelta(seconds=self.rollup)
+        while start < end:
+            yield start
+            start += delta
+
+
+class MockDataSource:
+
+    _tags = {
+        "environment": [
+            "production",
+            "staging",
+        ],
+        "release": [  # High cardinality
+            f"{major}.{minor}.{bugfix}"
```

Comment:
Let's make this more realistic by adding a a package:

```suggestion
            f"myapp@{major}.{minor}.{bugfix}"
```